File: content/snippets/python/design_patterns_bridge.py
from dataclasses import dataclass
from typing import Protocol


class Sender(Protocol):
    def send(self, message: str) -> None: ...


class EmailSender:
    def send(self, message: str) -> None:
        print(f"email: {message}")


class SmsSender:
    def send(self, message: str) -> None:
        print(f"sms: {message}")


@dataclass
class AlertNotification:
    sender: Sender

    def notify(self, message: str) -> None:
        self.sender.send(f"alert -> {message}")


AlertNotification(EmailSender()).notify("CPU high")
AlertNotification(SmsSender()).notify("CPU high")

File: content/snippets/cpp/design_patterns_bridge.cpp
#include <iostream>
#include <memory>
#include <string>

struct Sender {
    virtual ~Sender() = default;
    virtual void send(const std::string& message) const = 0;
};

struct EmailSender : Sender {
    void send(const std::string& message) const override {
        std::cout << "email: " << message << '\n';
    }
};

struct SmsSender : Sender {
    void send(const std::string& message) const override {
        std::cout << "sms: " << message << '\n';
    }
};

class Notification {
public:
    explicit Notification(const Sender& sender) : sender_(sender) {}
    virtual ~Notification() = default;
    virtual void notify(const std::string& message) const = 0;

protected:
    const Sender& sender_;
};

class AlertNotification : public Notification {
public:
    using Notification::Notification;

    void notify(const std::string& message) const override {
        sender_.send("alert -> " + message);
    }
};

int main() {
    EmailSender email;
    SmsSender sms;
    AlertNotification(email).notify("CPU high");
    AlertNotification(sms).notify("CPU high");
}
